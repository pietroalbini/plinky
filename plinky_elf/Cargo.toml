[package]
name = "plinky_elf"
edition = "2021"

[dependencies]
plinky_diagnostics.workspace = true
plinky_macros.workspace = true
plinky_utils.workspace = true

[dev-dependencies]
anyhow.workspace = true
insta.workspace = true
tempfile.workspace = true

[lints]
workspace = true
