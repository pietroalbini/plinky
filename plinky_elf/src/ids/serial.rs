#[macro_export]
macro_rules! new_serial_ids {
    (
        $ids_vis:vis $ids_ty:ident {
            type SectionId = $section_vis:vis $section_ty:ident;
            type SymbolId = $symbol_vis:vis $symbol_ty:ident;
            type StringId = $string_vis:vis $string_ty:ident;
        }
    ) => {
        use $crate::ids::convert::IdConversionMap;
        use $crate::ids::convert::ConvertibleElfIds;
        use $crate::ids::{ElfIds, ReprIdGetters, StringIdGetters};
        use $crate::{ElfObject, ElfSectionContent};

        #[derive(Copy, Clone, PartialEq, Eq, PartialOrd, Ord, Hash)]
        $section_vis struct $section_ty(usize);

        impl std::fmt::Debug for $section_ty {
            fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
                write!(f, "section#{}", self.0)
            }
        }

        impl ReprIdGetters for $section_ty {
            fn repr_id(&self) -> String {
                format!("{}", self.0)
            }
        }

        #[derive(Copy, Clone, PartialEq, Eq, PartialOrd, Ord, Hash)]
        $symbol_vis struct $symbol_ty(usize);

        impl std::fmt::Debug for $symbol_ty {
            fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
                write!(f, "symbol#{}", self.0)
            }
        }

        impl ReprIdGetters for $symbol_ty {
            fn repr_id(&self) -> String {
                format!("{}", self.0)
            }
        }

        #[derive(Copy, Clone, PartialEq, Eq, PartialOrd, Ord, Hash)]
        $string_vis struct $string_ty($section_ty, u32);

        impl $string_ty {
            $string_vis fn new(section: $section_ty, offset: u32) -> Self {
                Self(section, offset)
            }
        }

        impl std::fmt::Debug for $string_ty {
            fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
                write!(f, "{:?}:string#{}", self.0, self.1)
            }
        }

        impl StringIdGetters<$ids_ty> for $string_ty {
            fn section(&self) -> &$section_ty {
                &self.0
            }

            fn offset(&self) -> u32 {
                self.1
            }
        }

        #[derive(Debug)]
        $ids_vis struct $ids_ty {
            next_section_id: usize,
            next_symbol_id: usize,
        }

        impl ElfIds for $ids_ty {
            type SectionId = $section_ty;
            type SymbolId = $symbol_ty;
            type StringId = $string_ty;
        }

        impl<F> ConvertibleElfIds<F> for $ids_ty
        where
            F: ElfIds,
            F::StringId: StringIdGetters<F>,
        {
            fn create_conversion_map(
                &mut self,
                object: &ElfObject<F>,
                string_ids: &[F::StringId],
            ) -> IdConversionMap<F, Self> {
                let mut map = IdConversionMap::<F, Self>::new();

                for (old_id, section) in &object.sections {
                    map.section_ids.insert(old_id.clone(), self.allocate_section_id());

                    if let ElfSectionContent::SymbolTable(table) = &section.content {
                        for id in table.symbols.keys() {
                            map.symbol_ids.insert(id.clone(), self.allocate_symbol_id());
                        }
                    }
                }

                for string_id in string_ids {
                    map.string_ids.insert(
                        string_id.clone(),
                        $string_ty::new(
                            *map.section_ids.get(string_id.section()).expect("missing section"),
                            string_id.offset(),
                        ),
                    );
                }

                map
            }
        }

        impl $ids_ty {
            $ids_vis fn new() -> Self {
                Self { next_section_id: 0, next_symbol_id: 0 }
            }

            $ids_vis fn allocate_section_id(&mut self) -> $section_ty {
                let id = $section_ty(self.next_section_id);
                self.next_section_id += 1;
                id
            }

            $ids_vis fn allocate_symbol_id(&mut self) -> $symbol_ty {
                let id = $symbol_ty(self.next_symbol_id);
                self.next_symbol_id += 1;
                id
            }
        }
    };
}

new_serial_ids! {
    pub SerialIds {
        type SectionId = pub SerialSectionId;
        type SymbolId = pub SerialSymbolId;
        type StringId = pub SerialStringId;
    }
}
