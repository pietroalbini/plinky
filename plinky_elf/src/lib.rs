#![feature(error_generic_member_access)]

pub mod errors;
pub mod ids;
pub mod raw;
mod reader;
pub mod render_elf;
mod types;
pub mod writer;

pub use self::reader::{ReadSeek, ElfReader};
pub use self::types::*;
