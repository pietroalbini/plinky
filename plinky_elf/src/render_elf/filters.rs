use plinky_macros::{Display, Error};
use plinky_utils::filters_parser::{FilterParseError, FilterPart, StringFilter};

#[derive(Debug, Clone, PartialEq, Eq, PartialOrd, Ord)]
pub struct RenderElfFilters {
    pub(super) meta: bool,
    sections: SectionsFilter,
    pub(super) segments: bool,
}

impl RenderElfFilters {
    pub fn all() -> Self {
        RenderElfFilters { meta: true, sections: SectionsFilter::All, segments: true }
    }

    pub fn parse(raw: &str) -> Result<Self, RenderElfFiltersParseError> {
        let mut filter =
            RenderElfFilters { meta: false, sections: SectionsFilter::None, segments: false };

        for part in FilterPart::parse_iter(raw) {
            match part? {
                FilterPart::Special("meta") => filter.meta = true,
                FilterPart::Special("segments") => filter.segments = true,
                FilterPart::Special("sections") => match &filter.sections {
                    SectionsFilter::Some(_) => {
                        return Err(RenderElfFiltersParseError::CantMixSectionFilters)
                    }
                    SectionsFilter::None | SectionsFilter::All => {
                        filter.sections = SectionsFilter::All;
                    }
                },
                FilterPart::Special(other) => {
                    return Err(RenderElfFiltersParseError::InvalidSpecialFilter(other.into()))
                }
                FilterPart::StringFilter(section_filter) => match &mut filter.sections {
                    SectionsFilter::Some(section_filters) => section_filters.push(section_filter),
                    SectionsFilter::All => {
                        return Err(RenderElfFiltersParseError::CantMixSectionFilters)
                    }
                    SectionsFilter::None => {
                        filter.sections = SectionsFilter::Some(vec![section_filter])
                    }
                },
            }
        }

        Ok(filter)
    }

    pub(super) fn section(&self, name: &str) -> bool {
        match &self.sections {
            SectionsFilter::None => false,
            SectionsFilter::All => true,
            SectionsFilter::Some(filters) => filters.iter().any(|filter| filter.matches(name)),
        }
    }
}

#[derive(Debug, Clone, PartialEq, Eq, PartialOrd, Ord)]
enum SectionsFilter {
    None,
    All,
    Some(Vec<StringFilter>),
}

#[derive(Debug, Display, Error, PartialEq, Eq)]
pub enum RenderElfFiltersParseError {
    #[display("invalid special filter: {f0}")]
    InvalidSpecialFilter(String),
    #[display("can't mix @sections and individual section filters")]
    CantMixSectionFilters,
    #[transparent]
    Utils(FilterParseError),
}
