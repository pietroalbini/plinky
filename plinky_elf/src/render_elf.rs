use crate::ids::serial::{SectionId, SerialIds};
use crate::ids::StringIdGetters;
use crate::{
    ElfABI, ElfClass, ElfEndian, ElfMachine, ElfObject, ElfPermissions, ElfSectionContent,
    ElfSegmentContent, ElfSegmentType, ElfType,
};
use plinky_diagnostics::widgets::{Table, Text, Widget};

pub fn render_elf(object: &ElfObject<SerialIds>) -> impl Widget {
    let widgets: Vec<Box<dyn Widget>> = vec![
        Box::new(render_meta(object)),
        Box::new(render_sections(object)),
        Box::new(render_segments(object)),
    ];
    widgets
}

fn render_meta(object: &ElfObject<SerialIds>) -> impl Widget {
    let mut table = Table::new();
    table.set_title("Metadata:");
    table.add_row([
        "Class",
        match object.env.class {
            ElfClass::Elf32 => "ELF 32bit",
            ElfClass::Elf64 => "ELF 64bit",
        },
    ]);
    table.add_row([
        "Endian",
        match object.env.endian {
            ElfEndian::Little => "Little",
        },
    ]);
    table.add_row([
        "ABI",
        match object.env.abi {
            ElfABI::SystemV => "System V",
        },
    ]);
    table.add_row([
        "Machine",
        match object.env.machine {
            ElfMachine::X86 => "x86",
            ElfMachine::X86_64 => "x86-64",
        },
    ]);
    table.add_row([
        "Type",
        match object.type_ {
            ElfType::Relocatable => "Relocatable",
            ElfType::Executable => "Executable",
            ElfType::SharedObject => "Shared object",
            ElfType::Core => "Core dump",
        },
    ]);
    table.add_row([
        "Entrypoint".to_string(),
        match object.entry {
            Some(entry) => entry.to_string(),
            None => "-".to_string(),
        },
    ]);
    table
}

fn render_sections(object: &ElfObject<SerialIds>) -> impl Widget {
    Text::new(format!("{:#?}", object.sections))
}

fn render_segments(object: &ElfObject<SerialIds>) -> impl Widget {
    let mut table = Table::new();
    table.set_title("Segments:");
    table.add_row(["Type", "Perms", "Aligment", "Content"]);
    for segment in &object.segments {
        table.add_row([
            match segment.type_ {
                ElfSegmentType::Null => "Null".into(),
                ElfSegmentType::Load => "Load".into(),
                ElfSegmentType::Dynamic => "Dynamic".into(),
                ElfSegmentType::Interpreter => "Interpreter".into(),
                ElfSegmentType::Note => "Note".into(),
                ElfSegmentType::ProgramHeaderTable => "Program header table".into(),
                ElfSegmentType::GnuStack => "GNU stack".into(),
                ElfSegmentType::Unknown(id) => format!("<unknown: {id:#x}>"),
            },
            render_perms(&segment.perms),
            format!("{:#x}", segment.align),
            match &segment.content {
                ElfSegmentContent::Empty => "-".into(),
                ElfSegmentContent::Sections(sections) => sections
                    .iter()
                    .map(|&id| section_name(object, id))
                    .collect::<Vec<_>>()
                    .join(", "),
                ElfSegmentContent::Unknown(unknown) => format!(
                    "file: {:#x} (len: {:#x}), memory: {:#x} (len: {:#x})",
                    unknown.file_offset,
                    unknown.file_size,
                    unknown.virtual_address,
                    unknown.memory_size
                ),
            },
        ]);
    }
    table
}

fn render_perms(perms: &ElfPermissions) -> String {
    let mut output = String::new();
    let mut push = |cond: bool, chr: char| output.push(cond.then(|| chr).unwrap_or(' '));

    push(perms.read, 'R');
    push(perms.write, 'W');
    push(perms.execute, 'X');

    output
}

fn section_name(object: &ElfObject<SerialIds>, id: SectionId) -> String {
    let section = object.sections.get(&id).expect("invalid section id");
    let shstrtab = object.sections.get(&section.name.section()).expect("invalid string section id");
    let ElfSectionContent::StringTable(shstrtab) = &shstrtab.content else {
        panic!("string section id is not a string table");
    };
    let name = shstrtab.get(section.name.offset()).expect("missing section name");
    format!("{name}#{}", id.idx())
}
