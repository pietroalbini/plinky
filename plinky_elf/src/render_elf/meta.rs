use crate::ids::serial::SerialIds;
use crate::{ElfABI, ElfClass, ElfEndian, ElfMachine, ElfObject, ElfType};
use plinky_diagnostics::widgets::{Table, Widget};

pub(super) fn render_meta(object: &ElfObject<SerialIds>) -> impl Widget {
    let mut table = Table::new();
    table.set_title("Metadata:");
    table.add_row([
        "Class",
        match object.env.class {
            ElfClass::Elf32 => "ELF 32bit",
            ElfClass::Elf64 => "ELF 64bit",
        },
    ]);
    table.add_row([
        "Endian",
        match object.env.endian {
            ElfEndian::Little => "Little",
        },
    ]);
    table.add_row([
        "ABI",
        match object.env.abi {
            ElfABI::SystemV => "System V",
        },
    ]);
    table.add_row([
        "Machine",
        match object.env.machine {
            ElfMachine::X86 => "x86",
            ElfMachine::X86_64 => "x86-64",
        },
    ]);
    table.add_row([
        "Type",
        match object.type_ {
            ElfType::Relocatable => "Relocatable",
            ElfType::Executable => "Executable",
            ElfType::SharedObject => "Shared object",
            ElfType::Core => "Core dump",
        },
    ]);
    table.add_row([
        "Entrypoint".to_string(),
        match object.entry {
            Some(entry) => format!("{entry:#x}"),
            None => "-".to_string(),
        },
    ]);
    table
}
