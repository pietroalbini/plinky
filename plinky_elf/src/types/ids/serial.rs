use super::convert::IdConversionMap;
use crate::ids::convert::ConvertibleElfIds;
use crate::ids::{ElfIds, ReprIdGetters, StringIdGetters};
use crate::{ElfObject, ElfSectionContent};

#[derive(Copy, Clone, PartialEq, Eq, PartialOrd, Ord, Hash)]
pub struct SectionId(usize);

impl std::fmt::Debug for SectionId {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        write!(f, "section#{}", self.0)
    }
}

impl ReprIdGetters for SectionId {
    fn repr_id(&self) -> String {
        format!("{}", self.0)
    }
}

#[derive(Copy, Clone, PartialEq, Eq, PartialOrd, Ord, Hash)]
pub struct SymbolId(usize);

impl std::fmt::Debug for SymbolId {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        write!(f, "symbol#{}", self.0)
    }
}

impl ReprIdGetters for SymbolId {
    fn repr_id(&self) -> String {
        format!("{}", self.0)
    }
}

#[derive(Copy, Clone, PartialEq, Eq, PartialOrd, Ord, Hash)]
pub struct StringId(SectionId, u32);

impl StringId {
    pub fn new(section: SectionId, offset: u32) -> Self {
        Self(section, offset)
    }
}

impl std::fmt::Debug for StringId {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        write!(f, "{:?}:string#{}", self.0, self.1)
    }
}

impl StringIdGetters<SerialIds> for StringId {
    fn section(&self) -> &SectionId {
        &self.0
    }

    fn offset(&self) -> u32 {
        self.1
    }
}

#[derive(Debug)]
pub struct SerialIds {
    next_section_id: usize,
    next_symbol_id: usize,
}

impl ElfIds for SerialIds {
    type SectionId = SectionId;
    type SymbolId = SymbolId;
    type StringId = StringId;
}

impl<F> ConvertibleElfIds<F> for SerialIds
where
    F: ElfIds,
    F::StringId: StringIdGetters<F>,
{
    fn create_conversion_map(
        &mut self,
        object: &ElfObject<F>,
        string_ids: &[F::StringId],
    ) -> IdConversionMap<F, Self> {
        let mut map = IdConversionMap::<F, Self>::new();

        for (old_id, section) in &object.sections {
            map.section_ids.insert(old_id.clone(), self.allocate_section_id());

            if let ElfSectionContent::SymbolTable(table) = &section.content {
                for id in table.symbols.keys() {
                    map.symbol_ids.insert(id.clone(), self.allocate_symbol_id());
                }
            }
        }

        for string_id in string_ids {
            map.string_ids.insert(
                string_id.clone(),
                StringId::new(
                    *map.section_ids.get(string_id.section()).expect("missing section"),
                    string_id.offset(),
                ),
            );
        }

        map
    }
}

impl SerialIds {
    pub fn new() -> Self {
        Self { next_section_id: 0, next_symbol_id: 0 }
    }

    pub fn allocate_section_id(&mut self) -> SectionId {
        let id = SectionId(self.next_section_id);
        self.next_section_id += 1;
        id
    }

    pub fn allocate_symbol_id(&mut self) -> SymbolId {
        let id = SymbolId(self.next_symbol_id);
        self.next_symbol_id += 1;
        id
    }
}
