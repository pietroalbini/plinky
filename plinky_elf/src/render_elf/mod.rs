use crate::ids::serial::SerialIds;
use crate::render_elf::utils::{render_perms, section_name, MultipleWidgets};
use crate::ElfObject;
use plinky_diagnostics::widgets::Widget;

mod meta;
mod sections;
mod segments;
mod utils;

pub fn render(object: &ElfObject<SerialIds>) -> impl Widget {
    let mut widgets: Vec<Box<dyn Widget>> = Vec::new();
    widgets.push(Box::new(meta::render_meta(object)));
    for (&id, section) in &object.sections {
        widgets.push(Box::new(sections::render_section(object, id, section)));
    }
    widgets.push(Box::new(segments::render_segments(object)));
    MultipleWidgets(widgets)
}
