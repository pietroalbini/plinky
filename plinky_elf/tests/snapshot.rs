//! Check whether ELF files are parsed correctly.

use anyhow::{Context, Error};
use plinky_diagnostics::widgets::Widget;
use plinky_elf::ids::serial::SerialIds;
use plinky_elf::render_elf::RenderElfFilters;
use plinky_elf::ElfObject;
use std::fs::File;
use std::io::BufReader;
use std::path::Path;
use std::process::Command;
use tempfile::NamedTempFile;

macro_rules! test {
    ($name:ident, $source:expr $(, $variant:ident)*$(,)?) => {
        mod $name {
            use super::*;

            $(
                #[test]
                #[allow(non_snake_case)]
                fn $variant() -> Result<(), Error> {
                    implement_test($source, stringify!($variant))
                }
            )*
        }
    };
}

test!(
    hello_asm,
    "tests/snapshot/hello_asm.asm",
    x86,
    x86_64,
    x86__linked,
    x86_64__linked,
    x86__written,
    x86_64__written,
);
test!(hello_c, "tests/snapshot/hello_c.c", x86, x86_64);

#[track_caller]
fn implement_test(source: &str, name: &str) -> Result<(), Error> {
    let meta = Metadata::from_name(name);

    let mut object_file = match source.rsplit_once('.').map(|(_name, ext)| ext) {
        Some("asm") => compile_asm(source, meta.variant)?,
        Some("c") => compile_c(source, meta.variant)?,
        Some(other) => panic!("unsupported extension: {other}"),
        None => panic!("missing extension for {source}"),
    };
    if let Link::Yes = meta.link {
        object_file = link_single_object(object_file.path(), meta.variant)?;
    }
    let mut file = BufReader::new(File::open(object_file.path())?);

    let mut parsed = ElfObject::load(&mut file, &mut SerialIds::new())?;

    let mut settings = insta::Settings::clone_current();
    settings.set_omit_expression(true);
    settings.set_snapshot_path("snapshot");
    settings.set_prepend_module_to_snapshot(false);
    settings.set_snapshot_suffix(meta.snapshot_suffix());
    let _guard = settings.bind_to_scope();

    let name = source.rsplit_once('/').map(|(_dir, name)| name).unwrap_or(source);
    let name = name.rsplit_once('.').map(|(name, _ext)| name).unwrap_or(name);

    match meta.mode {
        Mode::Read => {}
        Mode::WriteThenRead => {
            let mut buf = Vec::new();
            parsed.write(&mut buf).context("failed to write back the ELF file")?;
            parsed = ElfObject::load(&mut std::io::Cursor::new(&mut buf), &mut SerialIds::new())?;
        }
    }

    let rendered =
        plinky_elf::render_elf::render(&parsed, &RenderElfFilters::all()).render_to_string();
    insta::assert_snapshot!(name, rendered);
    Ok(())
}

fn compile_asm(source: &str, variant: Variant) -> Result<NamedTempFile, Error> {
    let dest = NamedTempFile::new()?;
    let status = Command::new("nasm")
        .arg("-o")
        .arg(dest.path())
        .arg("-f")
        .arg(match variant {
            Variant::X86 => "elf32",
            Variant::X86_64 => "elf64",
        })
        .arg(source)
        .status()?;
    anyhow::ensure!(status.success(), "failed to compile {}", source);
    Ok(dest)
}

fn compile_c(source: &str, variant: Variant) -> Result<NamedTempFile, Error> {
    let dest = NamedTempFile::new()?;
    let status = Command::new("gcc")
        .arg("-c")
        .arg("-o")
        .arg(dest.path())
        .arg(source)
        .arg("-fno-pic")
        .arg(match variant {
            Variant::X86 => "-m32",
            Variant::X86_64 => "-m64",
        })
        .status()?;
    anyhow::ensure!(status.success(), "failed to compile {}", source);
    Ok(dest)
}

fn link_single_object(object: &Path, variant: Variant) -> Result<NamedTempFile, Error> {
    let dest = NamedTempFile::new()?;
    let status = Command::new("ld")
        .arg("-o")
        .arg(dest.path())
        .arg(object)
        .arg("-znoexecstack")
        .args(match variant {
            Variant::X86 => ["-m", "elf_i386"],
            Variant::X86_64 => ["-m", "elf_x86_64"],
        })
        .status()?;
    anyhow::ensure!(status.success(), "failed to link {object:?}");
    Ok(dest)
}

#[derive(Clone, Copy)]
struct Metadata {
    variant: Variant,
    link: Link,
    mode: Mode,
}

impl Metadata {
    fn from_name(name: &str) -> Metadata {
        fn set<T>(component: &str, store: &mut Option<T>, value: T) {
            if store.is_some() {
                panic!("duplicate {component} in name");
            }
            *store = Some(value);
        }

        let mut variant = None;
        let mut link = None;
        let mut mode = None;

        for component in name.split("__") {
            match component {
                "x86" => set(component, &mut variant, Variant::X86),
                "x86_64" => set(component, &mut variant, Variant::X86_64),
                "linked" => set(component, &mut link, Link::Yes),
                "written" => set(component, &mut mode, Mode::WriteThenRead),
                other => panic!("unknown component {other}"),
            }
        }

        Metadata {
            variant: variant.expect("missing variant"),
            link: link.unwrap_or(Link::No),
            mode: mode.unwrap_or(Mode::Read),
        }
    }

    fn snapshot_suffix(&self) -> String {
        let mut suffix = match self.variant {
            Variant::X86 => "32bit".to_string(),
            Variant::X86_64 => "64bit".to_string(),
        };
        if let Link::Yes = self.link {
            suffix.push_str("-linked");
        }
        if let Mode::WriteThenRead = self.mode {
            suffix.push_str("-written");
        }
        suffix
    }
}

#[derive(Clone, Copy)]
enum Variant {
    X86,
    X86_64,
}

#[derive(Clone, Copy)]
enum Link {
    Yes,
    No,
}

#[derive(Clone, Copy)]
enum Mode {
    Read,
    WriteThenRead,
}
