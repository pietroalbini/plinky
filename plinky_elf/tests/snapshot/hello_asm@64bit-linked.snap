---
source: plinky_elf/tests/snapshot.rs
---
  Metadata:
╭────────────┬────────────╮
│ Class      │ ELF 64bit  │
├────────────┼────────────┤
│ Endian     │ Little     │
├────────────┼────────────┤
│ ABI        │ System V   │
├────────────┼────────────┤
│ Machine    │ x86-64     │
├────────────┼────────────┤
│ Type       │ Executable │
├────────────┼────────────┤
│ Entrypoint │ 0x401000   │
╰────────────┴────────────╯

section #0 (address: 0x0)
 │
 │  empty section
 ┴

section .text#1 (address: 0x401000)
 │
 │  program data | permissions: R X
 │
 │  ╭─────────────────────────────────────────────────┬──────────────────╮
 │  │ b8 04 00 00 00 bb 01 00 00 00 b9 00 20 40 00 ba │ ............ @.. │
 │  │ 0d 00 00 00 cd 80 b8 04 00 00 00 bb 01 00 00 00 │ ................ │
 │  │ b9 0d 20 40 00 ba 0f 00 00 00 cd 80 b0 01 bb 00 │ .. @............ │
 │  │ 00 00 00 cd 80                                  │ .....            │
 │  ╰─────────────────────────────────────────────────┴──────────────────╯
 ┴

section .data#2 (address: 0x402000)
 │
 │  program data | permissions: RW
 │
 │  ╭─────────────────────────────────────────────────┬──────────────────╮
 │  │ 48 65 6c 6c 6f 20 77 6f 72 6c 64 21 0a 47 6f 6f │ Hello world!.Goo │
 │  │ 64 62 79 65 20 77 6f 72 6c 64 21 0a             │ dbye world!.     │
 │  ╰─────────────────────────────────────────────────┴──────────────────╯
 ┴

section .bss#3 (address: 0x402020)
 │
 │  uninitialized | len: 0x8 | permissions: RW 
 ┴

section .symtab#4 (address: 0x0)
 │
 │    Symbol table:
 │  ╭───────────────┬─────────┬────────┬────────────┬──────────┬──────╮
 │  │ Name          │ Binding │ Type   │ Definition │ Value    │ Size │
 │  ├───────────────┼─────────┼────────┼────────────┼──────────┼──────┤
 │  │ #0            │ Local   │ -      │ Undefined  │ 0x0      │ 0x0  │
 │  ├───────────────┼─────────┼────────┼────────────┼──────────┼──────┤
 │  │ hello_asm.S#1 │ Local   │ File   │ Absolute   │ 0x0      │ 0x0  │
 │  ├───────────────┼─────────┼────────┼────────────┼──────────┼──────┤
 │  │ msg_hello#2   │ Local   │ -      │ .data#2    │ 0x402000 │ 0x0  │
 │  ├───────────────┼─────────┼────────┼────────────┼──────────┼──────┤
 │  │ len_hello#3   │ Local   │ -      │ Absolute   │ 0xd      │ 0x0  │
 │  ├───────────────┼─────────┼────────┼────────────┼──────────┼──────┤
 │  │ msg_goodbye#4 │ Local   │ -      │ .data#2    │ 0x40200d │ 0x0  │
 │  ├───────────────┼─────────┼────────┼────────────┼──────────┼──────┤
 │  │ len_goodbye#5 │ Local   │ -      │ Absolute   │ 0xf      │ 0x0  │
 │  ├───────────────┼─────────┼────────┼────────────┼──────────┼──────┤
 │  │ uninit#6      │ Local   │ Object │ .bss#3     │ 0x402020 │ 0x8  │
 │  ├───────────────┼─────────┼────────┼────────────┼──────────┼──────┤
 │  │ _start#7      │ Global  │ -      │ .text#1    │ 0x401000 │ 0x0  │
 │  ├───────────────┼─────────┼────────┼────────────┼──────────┼──────┤
 │  │ __bss_start#8 │ Global  │ -      │ .bss#3     │ 0x40201c │ 0x0  │
 │  ├───────────────┼─────────┼────────┼────────────┼──────────┼──────┤
 │  │ _edata#9      │ Global  │ -      │ .data#2    │ 0x40201c │ 0x0  │
 │  ├───────────────┼─────────┼────────┼────────────┼──────────┼──────┤
 │  │ _end#10       │ Global  │ -      │ .bss#3     │ 0x402028 │ 0x0  │
 │  ╰───────────────┴─────────┴────────┴────────────┴──────────┴──────╯
 ┴

section .strtab#5 (address: 0x0)
 │
 │    Strings table:
 │  ╭──────┬─────────────╮
 │  │ 0x0  │             │
 │  ├──────┼─────────────┤
 │  │ 0x1  │ hello_asm.S │
 │  ├──────┼─────────────┤
 │  │ 0xd  │ msg_hello   │
 │  ├──────┼─────────────┤
 │  │ 0x17 │ len_hello   │
 │  ├──────┼─────────────┤
 │  │ 0x21 │ msg_goodbye │
 │  ├──────┼─────────────┤
 │  │ 0x2d │ len_goodbye │
 │  ├──────┼─────────────┤
 │  │ 0x39 │ uninit      │
 │  ├──────┼─────────────┤
 │  │ 0x40 │ __bss_start │
 │  ├──────┼─────────────┤
 │  │ 0x4c │ _edata      │
 │  ├──────┼─────────────┤
 │  │ 0x53 │ _end        │
 │  ╰──────┴─────────────╯
 ┴

section .shstrtab#6 (address: 0x0)
 │
 │    Strings table:
 │  ╭──────┬───────────╮
 │  │ 0x0  │           │
 │  ├──────┼───────────┤
 │  │ 0x1  │ .symtab   │
 │  ├──────┼───────────┤
 │  │ 0x9  │ .strtab   │
 │  ├──────┼───────────┤
 │  │ 0x11 │ .shstrtab │
 │  ├──────┼───────────┤
 │  │ 0x1b │ .text     │
 │  ├──────┼───────────┤
 │  │ 0x21 │ .data     │
 │  ├──────┼───────────┤
 │  │ 0x27 │ .bss      │
 │  ╰──────┴───────────╯
 ┴

  Segments:
╭───────────┬───────┬──────────┬───────────────────────────────────────────────────────╮
│ Type      │ Perms │ Aligment │ Content                                               │
├───────────┼───────┼──────────┼───────────────────────────────────────────────────────┤
│ Load      │ R     │ 0x1000   │ file: 0x0 (len: 0x120), memory: 0x400000 (len: 0x120) │
├───────────┼───────┼──────────┼───────────────────────────────────────────────────────┤
│ Load      │ R X   │ 0x1000   │ .text#1                                               │
├───────────┼───────┼──────────┼───────────────────────────────────────────────────────┤
│ Load      │ RW    │ 0x1000   │ .data#2                                               │
├───────────┼───────┼──────────┼───────────────────────────────────────────────────────┤
│ GNU stack │ RW    │ 0x10     │ -                                                     │
╰───────────┴───────┴──────────┴───────────────────────────────────────────────────────╯
